### ISSUE TITLE
Time-independent acceptance criterion

### ISSUE THREAD

**Issue by leonlan** [PyVRP/PyVRP#946]
Part of #533. The current acceptance criterion is bound to `StoppingCriterion.fraction_remaining`. When using MaxRuntime as criterion, we can't reproduce the runs because it's tightly coupled to the runtime per iteration.

We need something that's independent from the runtime criteria, but works just as well.

**Comment by leonlan** [comment]
The main challenge is that every instance has some kind of 'sweet spot' for $\alpha \in [0, 1]$ in the acceptance criterion. I observed that prize-collecting instances benefit from higher $\alpha$ whereas CVRP instances benefit from lower $\alpha$. Using a MaxRuntime criterion and using its fraction remaining performs best for all scenarios on average. 

Can we instead somehow dynamically adjust $\alpha$? I need to make some plots of the search trajectories to get a clear picture of this. 


What's strange about this $\alpha$ problem is that shorter runtime sessions consistently outperform the intermediate checkpoints of longer sessions. For example, a dedicated 15s run achieves a solution of `73936`, while the 15s checkpoint within a 30s run only reaches `74380`. The same pattern appears when comparing 30s runs: a standalone 30s run finds `73408`, whereas the 30s mark in a 60s session yields `73785`.


<details>
<summary>15s runtime</summary>


```
    Iters    Time |      Current OK    Candidate OK         Best OK
H     926      1s |        78499  Y        78499  Y        78161  Y
H    1895      2s |        77447  Y        77447  Y        77294  Y
H    2941      3s |        77215  Y        77414  Y        76959  Y
H    4185      4s |        76793  Y        77088  Y        76618  Y
H    5380      5s |        76274  Y        76274  Y        76217  Y
H    6653      6s |        76026  Y        76026  Y        75926  Y
H    7944      7s |        75663  Y        75832  Y        75593  Y
H    9255      8s |        75240  Y        75240  Y        75213  Y
H   10521      9s |        75040  Y        75040  Y        75010  Y
H   11847     10s |        74778  Y        75019  N        74711  Y
H   13270     11s |        74358  Y        74405  Y        74341  Y
H   14688     12s |        74186  Y        74727  Y        74176  Y
H   16069     13s |        74036  Y        74193  N        74026  Y
H   17533     14s |        73942  Y        73995  Y        73936  Y
```

</details>


<details>
<summary>30s runtime</summary>

```
    Iters    Time |      Current OK    Candidate OK         Best OK
H     996      1s |        78367  Y        78367  Y        77951  Y
H    2001      2s |        77810  Y        77810  Y        77293  Y
H    3081      3s |        77257  Y        77257  Y        76969  Y
H    4127      4s |        76874  Y        76874  Y        76573  Y
H    5055      5s |        76483  Y        76483  Y        76275  Y
H    6193      6s |        76399  Y        76464  Y        76250  Y
H    7305      7s |        76128  Y        76272  Y        76049  Y
H    8517      8s |        75938  Y        75938  Y        75715  Y
H    9743      9s |        75603  N        75603  N        75494  Y
H   10887     10s |        75415  Y        75415  Y        75309  Y
H   12241     11s |        75192  Y        75321  Y        75105  Y
H   13697     12s |        74959  Y        75043  Y        74872  Y
H   15075     13s |        74719  Y        74901  Y        74622  Y
H   16474     14s |        74536  Y        74557  Y        74472  Y
H   17884     15s |        74448  Y        74448  Y        74380  Y <---- 15s
H   19303     16s |        74347  Y        74465  N        74267  Y
H   20604     17s |        74177  Y        74252  N        74128  Y
H   21999     18s |        74030  Y        74151  Y        73994  Y
H   23404     19s |        73992  Y        73992  Y        73954  Y
H   24943     20s |        73897  Y        73944  Y        73860  Y
H   26537     21s |        73786  Y        73863  Y        73752  Y
H   28078     22s |        73718  Y        73940  Y        73700  Y
H   29628     23s |        73651  Y        73826  Y        73636  Y
H   31235     24s |        73591  Y        73602  Y        73572  Y
H   32746     25s |        73563  Y        73725  N        73549  Y
H   34132     26s |        73478  Y        73478  Y        73465  Y
H   35595     27s |        73450  Y        73470  Y        73442  Y
H   37189     28s |        73430  Y        73510  Y        73425  Y
H   38674     29s |        73411  Y        73647  Y        73408  Y
```

</details>


<details>
<summary>60s runtime</summary>

```
    Iters    Time |      Current OK    Candidate OK         Best OK
H    1019      1s |        78371  Y        78465  Y        77824  Y
H    2068      2s |        77683  Y        77683  Y        77375  Y
H    3239      3s |        77168  Y        77168  Y        76824  Y
     4440      4s |        76899  Y        76899  Y        76824  Y
H    5715      5s |        76722  N        76722  N        76463  Y
H    6975      6s |        76490  Y        76490  Y        76250  Y
H    8301      7s |        76292  Y        76423  Y        76104  Y
H    9651      8s |        76180  Y        76344  Y        75989  Y
H   10989      9s |        76030  Y        76030  Y        75899  Y
H   12416     10s |        75833  Y        75833  Y        75814  Y
H   13770     11s |        75763  Y        75763  Y        75639  Y
H   15191     12s |        75458  Y        75618  Y        75281  Y
H   16595     13s |        75230  Y        75360  Y        75168  Y
H   17913     14s |        75194  Y        75194  Y        75054  Y
H   19258     15s |        75052  Y        75184  Y        74935  Y
H   20580     16s |        75023  Y        75098  Y        74917  Y
H   21963     17s |        74966  Y        74966  Y        74884  Y
H   23339     18s |        74822  Y        74977  Y        74747  Y
H   24805     19s |        74790  Y        74817  Y        74683  Y
H   26313     20s |        74713  Y        74713  Y        74620  Y
H   27860     21s |        74597  Y        74597  Y        74527  Y
H   29341     22s |        74509  Y        74637  Y        74462  Y
H   30854     23s |        74482  Y        74671  Y        74406  Y
H   32415     24s |        74363  Y        74363  Y        74341  Y
H   33961     25s |        74271  Y        74271  Y        74209  Y
H   35455     26s |        74271  Y        74474  Y        74189  Y
H   36960     27s |        74141  Y        74390  N        74072  Y
H   38500     28s |        73979  Y        74011  Y        73937  Y
H   40054     29s |        73867  Y        73913  Y        73815  Y
H   41548     30s |        73837  Y        73837  Y        73785  Y <---- 30s
H   43034     31s |        73790  Y        73945  N        73737  Y
H   44531     32s |        73773  Y        73773  Y        73724  Y
H   46017     33s |        73601  Y        73728  Y        73535  Y
H   47552     34s |        73541  Y        73901  Y        73507  Y
H   49143     35s |        73516  Y        73516  Y        73477  Y
H   50750     36s |        73473  Y        73536  Y        73426  Y
    52395     37s |        73473  Y        73500  Y        73426  Y
    54050     38s |        73489  Y        73556  Y        73426  Y
H   55599     39s |        73441  Y        73454  Y        73402  Y
H   57250     40s |        73409  Y        73455  Y        73388  Y
H   58920     41s |        73400  Y        73534  Y        73379  Y
H   60238     42s |        73331  Y        73331  Y        73306  Y
H   61973     43s |        73286  Y        73428  Y        73261  Y
H   63643     44s |        73238  Y        73431  Y        73220  Y
H   65325     45s |        73199  Y        73390  Y        73168  Y
    66926     46s |        73197  Y        73197  Y        73168  Y
    68485     47s |        73200  Y        73215  Y        73168  Y
H   70002     48s |        73120  Y        73330  Y        73118  Y
H   71590     49s |        73077  Y        73436  Y        73068  Y
H   73204     50s |        73077  Y        73131  Y        73062  Y
H   74854     51s |        73068  Y        73068  Y        73053  Y
H   76575     52s |        73060  Y        73070  Y        73050  Y
H   78208     53s |        73039  Y        73120  N        73033  Y
H   79901     54s |        73008  Y        73008  Y        73002  Y
H   81507     55s |        72988  Y        73032  Y        72988  Y
H   83197     56s |        72979  Y        73025  Y        72977  Y
H   84864     57s |        72930  Y        72930  Y        72926  Y
H   86538     58s |        72925  Y        72993  N        72925  Y
H   88231     59s |        72921  Y        73069  Y        72921  Y
```

</details>


<img width="1359" height="967" alt="Image" src="https://github.com/user-attachments/assets/e99aef91-914c-469d-a3fd-063f49a8b16d" />

**Comment by leonlan** [comment]
We could also use `iters_no_improvement` as fraction remaining, that is, `1 - iters_no_improvement/num_iters_no_improvement`. This idea is simple: if we haven't found any improvements for a while, then we lower the threshold. 

The downside is that the fraction remaining graph continuously moves back to `fraction_remaining = 1`. This is not good, because a weight of 1 often makes the current solution diverge away from the best-found solution. 

<img width="1447" height="806" alt="Image" src="https://github.com/user-attachments/assets/d11930d3-6079-4101-8020-fb4cb1232a26" />

But it's definitely an idea we can work out further. There are two directions that I can think of:

- When a new improvement is found:
	1. We "freeze" fraction remaining for some number of iterations. 
	2. We multiply the current fraction remaining by some factor, going up a bit but not by too much.
- When a restart happens:
	1. Instead of going back to 1, we go back to 0.5, 0.25, 0.125, 0.0625, etc. following a geometric series.

It could be helpful to look at this from the perspective of having a "budget" that renews/replenishes.


It's also important that the first search trajectory is not that important - it's a lot of quick gains.



-------------

Based on MaxIterations resetting every 20K

<img width="1415" height="782" alt="Image" src="https://github.com/user-attachments/assets/63cfa784-2b11-453c-8138-0743b997fa61" />

**Comment by leonlan** [comment]
There may be some relevant ideas in [this paper](https://www.sciencedirect.com/science/article/pii/S0305054818303277).

> Franzin, A., & Stützle, T. (2019). Revisiting simulated annealing: A component-based analysis. Computers & Operations Research, 104, 191–206. https://doi.org/10.1016/j.cor.2018.12.015

While not exactly similar, it might be interesting to look at how the ML community addresses the exploration/exploitation trade-off during training. The learning rate is somewhat analogous to our acceptance weight.

**Comment by leonlan** [comment]
A quick sanity check: what if we just perform hill-climbing, that is, the weight is always set to zero? Benchmark results:

| Commit  | Time     | CVRP-XXL | MDVRPTW | HFVRP | VRPB | MTVRPTWR | CVRP | PCVRPTW | VRPTW |
|---------|----------|----------|---------|-------|------|----------|------|---------|-------|
| bbdc29b (main)  | 1 min    | 4.37%    | 1.87%   | 1.61%  | 1.20%  | 1.00%    | 0.63%  | 0.93%   | 1.77%  |
|          | 2 min    | 3.50%    | 1.40%   | 1.10%  | 0.91%  | 0.73%    | 0.49%  | 0.73%   | 1.50%  |
|          | 10 min   | 1.91%    | 0.81%   | 0.58%  | 0.57%  | 0.32%    | 0.30%  | 0.43%   | 0.97%  |
| f5944da (hill-climbing) | 1min     |    4.60% |   3.33% | 2.67% | 1.77% |    1.11% | 0.91% |   2.59% | 3.50% |
|   | 2min     |    3.97% |   2.76% | 1.95% | 1.52% |    0.90% | 0.80% |   2.28% | 2.32% |
| | 10min    |    2.85% |   2.06% | 1.19% | 1.21% |    0.46% | 0.59% |   1.45% | 1.75% |

As expected, the hill climbing approach gets stuck in local optima and has no way to get out easily. The acceptance criterion is another way, besides perturbation, to get of local optima.


---

Continuing this small experiment: what if we fix the acceptance weight? Tried this for 0, 0.05, 0.10, 0.20, 0.25 and 0.50. Dashed line corresponds to main/the decaying version.

<img width="1863" height="984" alt="Image" src="https://github.com/user-attachments/assets/56b8c0b1-9964-47e6-95ba-322c5bdbbca6" />

**Comment by N-Wouda** [comment]
OK so weight ~0.15 seems best, if it's left fixed. But decaying is better in all cases. I suppose this is good enough for tuning this particular parameter.

**Comment by leonlan** [comment]
<img width="1826" height="905" alt="Image" src="https://github.com/user-attachments/assets/92c0f126-78ad-493d-9dfd-422ddf2076d3" />

Maybe we can learn something from here...

**Comment by leonlan** [comment]
Improving the primal integral of the time-dependent approach is fairly straight forward. We can add an inner loop which creates an additional multiplicative factor for the `weight` (see https://github.com/PyVRP/PyVRP/pull/973).

<img width="2700" height="1500" alt="Image" src="https://github.com/user-attachments/assets/d4e9972c-54d2-4b64-91d7-09ccc6acd489" />

**Comment by leonlan** [comment]
**bbdc29b (main)**

| Time | CVRP-XXL | MDVRPTW | HFVRP | VRPB | MTVRPTWR | CVRP | PCVRPTW | VRPTW |
|------|----------|---------|-------|------|----------|------|---------|-------|
| 1 min | 4.37% | 1.87% | 1.61% | 1.20% | 1.00% | 0.63% | 0.93% | 1.77% |
| 10 min | 1.91% | 0.81% | 0.58% | 0.57% | 0.32% | 0.30% | 0.43% | 0.97% |
| academic | 0.99% | 0.48% | 0.49% | 0.43% | 0.31% | 0.26% | 0.22% | 0.63% |

## Experimental 

| Commit | Time | CVRP-XXL | MDVRPTW | HFVRP | VRPB | MTVRPTWR | CVRP | PCVRPTW | VRPTW |
|--------|------|----------|---------|-------|------|----------|------|---------|-------|
| **d39d17e** | 1 min | 5.06% | 2.11% | 1.78% | 1.35% | 1.10% | 0.70% | 1.02% | 1.90% |
| | 10 min | 2.05% | 0.99% | 0.77% | 0.70% | 0.38% | 0.36% | 0.68% | 1.13% |
| | academic | 1.69% | 0.82% | 0.69% | 0.62% | 0.38% | 0.34% | 0.47% | 0.86% |
| **edcd1b3** | 1 min | 5.01% | 2.32% | 1.94% | 1.49% | 0.98% | 0.77% | 1.03% | 2.13% |
| | 10 min | 2.35% | 1.06% | 0.80% | 0.73% | 0.27% | 0.36% | 0.56% | 1.21% |
| **2ee14a3** | 1 min | 5.22% | 2.42% | 2.09% | 1.63% | 1.18% | 0.82% | 1.06% | 2.31% |
| **274ca4c** | 1 min | 5.51% | 3.04% | 2.18% | 1.70% | 1.53% | 1.01% | 1.03% | 2.47% |
| | 10 min | 2.31% | 0.83% | 0.67% | 0.68% | 0.20% | 0.30% | 0.40% | 1.12% |
| **ec831fc** | 1 min | 5.19% | 2.38% | 1.75% | 1.23% | 0.98% | 0.69% | 0.84% | 2.38% |
| | 10 min | 2.33% | 1.40% | 0.99% | 0.88% | 0.55% | 0.50% | 0.57% | 1.27% |
| **dfc60ba** | 1 min | 5.66% | 3.01% | 1.95% | 1.45% | 1.03% | 0.82% | 1.23% | 2.50% |
| | 10 min | 3.75% | 1.11% | 0.90% | 0.69% | 0.46% | 0.43% | 0.56% | 1.22% |
| | academic | 2.06% | 0.73% | 0.75% | 0.58% | 0.46% | 0.38% | 0.36% | 0.83% |
| **1ae46c1** | 1 min | 5.82% | 5.29% | 2.40% | 1.67% | 1.54% | 1.16% | 1.45% | 3.50% |
| | 10 min | 3.39% | 1.03% | 0.72% | 0.62% | 0.42% | 0.38% | 0.44% | 1.05% |
| | academic | 1.18% | 0.88% | 0.66% | 0.57% | 0.42% | 0.37% | 0.34% | 0.85% |
| **a824b4b** | 1 min | 4.80% | 2.18% | 1.76% | 1.27% | 0.83% | 0.66% | 0.99% | 2.13% |
| | 10 min | 2.41% | 1.16% | 0.96% | 0.89% | 0.38% | 0.44% | 0.60% | 1.26% |
| | academic | 1.63% | 0.94% | 0.89% | 0.79% | 0.39% | 0.42% | 0.36% | 0.96% |

**Comment by leonlan** [comment]
I tried a lot of variations with thresholds (and population) and the overall conclusion is that a time-dependent acceptance criterion cannot be outperformed because it's optimised to perform best for all time limits. Time-dependent criteria can balance out the exploration-exploitation over all time limits, which is not something we can mimic easily without time limit information.

A short write-up of things I've tried:
- Adaptive weight for current criterion
	-  Hard to determine when weight needs to increase/decrease: this is the same as the 'reheating' problem in simulated annealing. Assuming we get stuck in a local optimum, how much do we need to raise the threshold?
	- Cyclic, budgeted approach going from 1 to 0 every so many iterations. #973. This works OK but I find it very annoying to look at.
- Using a subpopulation without generation size (always purge). This mimics a history but includes fitness (cost + diversity) calculations. This worked OK for https://github.com/PyVRP/PyVRP/commit/a824b4b8a1bfad64def4c2f38c056012c1b5dd31
- Late acceptance hill climbing. Slow convergence, as also reported in the paper, but works fairly well on longer runtimes. https://github.com/PyVRP/PyVRP/commit/1ae46c193deb2df668cc7b7f3d983a5fb4ad524e

I tried a lot of variations of each method, and these were some things that were hard to deal with:
- Allowing infeasible solutions is very useful since it gives us more search space, but penalised cost handling in thresholds is difficult because penalised cost evaluation gets outdated.
- Restarting with thresholds is a difficult problem: this is known as the reheating problem in simulated annealing. Increasing thresholds too much loses too much information, and increasing too little keeps us stuck in the local optimum.


Inspired by all my attempts, I've settled on an exponential moving average (EMA) approach in #975. It combines our current moving-average approach, and its similar to what the behavior of how late acceptance hill climbing works. In particular:
- EMA's recency bias is pretty intuitive in a single-trajectory algorithm.
- Only decreases thresholds, so it behaves like record-to-record (we always stick to the global best).
